### PR TITLE
add lsm to v1.2.x

### PR DESCRIPTION
**Pull request by joe-bowman**
## 1. Summary
Add LSM to v1.2.x

## 2.Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## 3. Implementation details

- add lsm types
- add lsm caps (limits) and getters and setters
- add governance methods for updating lsm caps
- add delegationaccountbalances callback to check for multiple denoms (not just staking denom) and handle flushing delegations
- add tests for deposittx and accountbalance callbacks
- add basic validation for existing message types

### CODE REVIEW

**Review by ThanhNhann — COMMENTED** [review]
I left some cmts and checking more

File: proto/quicksilver/lsm-types/v1/validator.proto [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  cosmos.staking.v1beta1.Description description = 7 [(gogoproto.nullable) = false];
+  // unbonding_height defines, if unbonding, the height at which this validator has begun unbonding.
+  int64 unbonding_height = 8 [(gogoproto.moretags) = "yaml:\"unbonding_height\""];
+  // unbonding_time defines, if unbonding, the min time for the validator to complete unbonding.
+  google.protobuf.Timestamp unbonding_time = 9
+      [(gogoproto.nullable) = false, (gogoproto.stdtime) = true, (gogoproto.moretags) = "yaml:\"unbonding_time\""];
+  // commission defines the commission parameters.
+  cosmos.staking.v1beta1.Commission commission = 10 [(gogoproto.nullable) = false];
+  // Deprecated: This field has been deprecated with LSM in favor of the validator bond
+  string min_self_delegation = 11 [
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply proto/quicksilver/lsm-types/v1/validator.proto:53]: We still need to be able to unmarshal non-lsm chain validator structs. The implementation here is able to do both lsm and non-lsm unmarshalling.

File: proto/quicksilver/lsm-types/v1/validator.proto [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    deprecated             = true,
+    (gogoproto.moretags)   = "yaml:\"min_self_delegation\"",
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable)   = false
+  ];
+  // strictly positive if this validator's unbonding has been stopped by external modules
+  int64 unbonding_on_hold_ref_count = 12;
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply proto/quicksilver/lsm-types/v1/validator.proto:60]: As above. We still need to be able to unmarshal non-lsm chain validator structs. The implementation here is able to do both lsm and non-lsm unmarshalling.

File: x/interchainstaking/keeper/callbacks.go [hunk 3]
```diff
@@ -533,7 +578,21 @@ func DelegationAccountBalanceCallback(k Keeper, ctx sdk.Context, args []byte, qu
...
 	zone.WithdrawalWaitgroup--
+
+	// set the zone amount.
+	balance := zone.DelegationAddress.Balance
+	if ok, _ := zone.DelegationAddress.Balance.Find(coin.Denom); !ok {
+		zone.DelegationAddress.Balance = zone.DelegationAddress.Balance.Add(coin)
+	} else {
+		for idx, i := range balance {
+			if coin.Denom == i.Denom {
+				zone.DelegationAddress.Balance[idx].Amount = coin.Amount
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/callbacks.go:591]: Basically we want to update the `DelegationAccount.Balance` value to match the host chain, but we can only query one denom at a time; so for each response, we set the `Amount` of that denom to be equal to `coin.Amount`.

Once we have done this for each denom, DelegationAccount.Balance will match the host chain. It feels inelegant, and if you have a better solution I'd gladly hear it!

**Review by ThanhNhann — COMMENTED** [review]
I have some cmt for this, overall it looks very good with the huge work @joe-bowman

File: x/interchainstaking/keeper/delegation.go [hunk 4]
```diff
@@ -218,21 +218,42 @@ func (k *Keeper) PrepareDelegationMessagesForShares(_ sdk.Context, zone *types.Z
...
+		return out
+	}
+
+	for _, val := range zone.Validators {
+		// validator bond max
+		maxBondShares := val.ValidatorBondShares.Mul(caps.ValidatorBondCap).Sub(val.LiquidShares)
+
+		// validator pc max
+		maxLiquidStakedShares := sdk.NewDecFromInt(val.VotingPower).Mul(caps.ValidatorCap).Sub(val.LiquidShares)
+		out[val.ValoperAddress] = sdkmath.MinInt(maxBondShares.TruncateInt(), maxLiquidStakedShares.TruncateInt())
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/delegation.go:241]: We want to find the lower of the two max values (we cannot exceed either of them).

File: x/interchainstaking/keeper/ibc_packet_handlers.go [hunk 5]
```diff
@@ -136,35 +136,37 @@ func (k *Keeper) HandleAcknowledgement(ctx sdk.Context, packet channeltypes.Pack
 				return err
 			}
 			continue
-		case "/cosmos.staking.v1beta1.MsgRedeemTokensforShares":
-			// TODO: handle this before LSM
+		case "/cosmos.staking.v1beta1.MsgRedeemTokensForShares":
```
[the remark this answers is not in the document]
Comment (joe-bowman) [reply x/interchainstaking/keeper/ibc_packet_handlers.go:139]: No - the value is correct per cosmos-sdk v0.45.16-ics-lsm is as follows:

https://github.com/cosmos/cosmos-sdk/blob/0af2f4da004cbea6414a8bad56e8bdcd45badf1e/x/staking/types/tx.pb.go#L1457

[the remark this answers is not in the document]
Comment (ThanhNhann) [reply x/interchainstaking/keeper/ibc_packet_handlers.go:139]: Its like @phamminh0811 tested the path above but it didn't work, can you recheck and confirm this @phamminh0811

[the remark this answers is not in the document]
Comment (phamminh0811) [reply x/interchainstaking/keeper/ibc_packet_handlers.go:139]: I saw that on the older version of [liquistaking module](https://github.com/iqlusioninc/liquidity-staking-module/blob/1421767ddca598253a81ffe3e70fa8084f255b9e/x/staking/types/tx.pb.go#L1208), the msg type start with liquidstaking, now on [this](https://github.com/quicksilver-zone/quicksilver/blob/9b7f568afa2e3583f26a3d2e3f1c7003439178e4/x/lsmtypes/types.pb.go#L1833), it become cosmos so everything will work well

**Review by ThanhNhann — APPROVED** [review]
Overall lgtm, just need to add some cmts for the logic in functions so it will be easy for us to check and maintain in the futute

**Review by anhductn2001 — APPROVED** [review]

**Comment by faddat** [comment]
I think we should consider 1.4 in lieu of this

**Comment by joe-bowman** [comment]
> I think we should consider 1.4 in lieu of this

Agreed. Closing.

**Pull request closed without merging**